graylog2-server/src/main/java/org/graylog/metrics/prometheus/mapping/PrometheusMappingFilesHandler.java

```diff
@@ -0,0 +1,155 @@
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.metrics.prometheus.mapping;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.io.Resources;
+import com.google.inject.assistedinject.Assisted;
+import io.prometheus.client.dropwizard.samplebuilder.MapperConfig;
+import org.graylog2.plugin.utilities.FileInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+public class PrometheusMappingFilesHandler {
+    public interface Factory {
+        PrometheusMappingFilesHandler create(URL coreResource,
+                                             @Assisted("coreMappingFile") Path coreMappingFile,
+                                             @Assisted("customMappingFile") Path customMappingFile);
+    }
+
+    private static final Logger LOG = LoggerFactory.getLogger(PrometheusMappingFilesHandler.class);
+
+    private final URL coreResource;
+    private final Path coreMappingFile;
+    private final Path customMappingFile;
+    private final PrometheusMappingConfigLoader mapperConfigLoader;
+
+    private final AtomicReference<FileInfo> coreMappingFileInfoRef = new AtomicReference<>(null);
+    private final AtomicReference<FileInfo> customMappingFileInfoRef = new AtomicReference<>(null);
+
+    @Inject
+    public PrometheusMappingFilesHandler(@Assisted URL coreResource,
+                                         @Assisted("coreMappingFile") Path coreMappingFile,
+                                         @Assisted("customMappingFile") Path customMappingFile,
+                                         PrometheusMappingConfigLoader mapperConfigLoader) {
+        this.coreResource = coreResource;
+        this.coreMappingFile = coreMappingFile.toAbsolutePath();
+        this.customMappingFile = customMappingFile.toAbsolutePath();
+        this.mapperConfigLoader = mapperConfigLoader;
+    }
+
+    private Optional<FileInfo.Change> detectChange(Path file, AtomicReference<FileInfo> fileInfoRef) {
+        if (Files.exists(file)) {
+            LOG.debug("Detecting changes for file <{}>", file);
+
+            if (fileInfoRef.get() == null) {
+                LOG.debug("Getting initial file info for file <{}>", file);
+                fileInfoRef.set(FileInfo.forPath(file));
+            }
+
+            final FileInfo.Change change = fileInfoRef.get().checkForChange();
+
+            if (change.isChanged()) {
```

Comment:
I think this check missing the edge case that the mappings (core or custom) file does not exist at server start time, then is added later.  Once the service starts, then the file is added later, the mappings are not loaded when the initial file info is loaded during refresh, so second change to the file is needed to force the change to be detected and the metrics in the file to be loaded.

Reply:
Good catch, thanks! Fixed in https://github.com/Graylog2/graylog2-server/pull/10767/commits/fb8be5536c2814fb7ddb304ee7bd0ea1347d3c41.